doc/smart-switch/smart-switch-database-architecture/smart-switch-database-design.md

```diff
@@ -0,0 +1,262 @@
+    "state": "always_enabled",
+    "support_syslog_rate_limit": "true"
+},
+```
+
+Within the NPU, the management of DPU overlay databases involves specific configurations. Each DPU overlay database instance is bound to the IP address of the midplane bridge (169.254.200.254 by default). The TCP port assignment follows a predictable pattern, with each DPU ID associated with a unique port (6380 + DPU ID). Additionally, the Unix domain socket path is structured as /var/run/redisdpu{DPU_ID}.
+
+Here is an example includes two DPU:
+
+``` json
+# DPU0: /var/run/redisdpu2/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6381,
+    "unix_socket_path": "/var/run/redisdpu0/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+#DPU1: /var/run/redisdpu1/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6382,
+    "unix_socket_path": "/var/run/redisdpu1/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+```
+
+There are four new tables introduction for the DPU overlay database:
+
+``` json
+"DPU_APPL_DB": {
+    "id": 15,
+    "separator": ":",
+    "instance": "redis",
+    "format": "proto"
+},
+"DPU_APPL_STATE_DB": {
+    "id": 16,
+    "separator": "|",
+    "instance": "redis"
+},
+"DPU_STATE_DB": {
+    "id": 17,
+    "separator": "|",
+    "instance": "redis"
+},
+"DPU_COUNTERS_DB": {
+    "id": 18,
+    "separator": ":",
+    "instance": "redis"
+}
+```
+
+#### Database flow
+
+This section outlines critical workflows interacting with the DPU overlay database.
+
+- Update Overlay Objects via GNMI:
+
+Communication with the SWSS of the DPU occurs through GNMI, leveraging ZMQ. Simultaneously, an asynchronous insertion of the object backup is made to the DPU_APPL_DB. This backup mechanism serves purposes such as debugging, migration, and future considerations.
+
+- Update Object Status:
+
+The SWSS of the DPU takes a proactive role in updating the DPU_APPL_STATE_DB and DPU_STATE_DB when corresponding objects undergo updates. This update can be triggered either by GNMI message commands or internal service logic.
+
+- Update Counters and Meters:
+
+Flex counter management in Syncd of the DPU handles the update of counters and meters for overlay objects. Traditional counters are also managed through this mechanism.
+
+These workflows ensure an interaction between the DPU overlay database and various components within the Smart Switch. The DPUs access their respective database instances via the IP address of the midplane bridge and the assigned TCP port. Concurrently, GNMI accesses these instances through the Unix domain socket
+
+### High-Level Design 
+
+### SAI API 
+
+N/A
+
+### Configuration and management 
+
+No new CLI commands are required
+
+#### CLI/YANG model Enhancements 
+
+
+``` yang
+    container sonic-feature {
+        container FEATURE {
+                leaf has_per_dpu_scope {
```

Comment:
is this feature used or expected? just to confirm, since I am not aware that we have another mode.

also, for objects inside of db, it will be better to add a link to maybe dash api proto repo.

Reply:
It just follows the multi-asic logic. 
IMO, This field does explicitly tell users that corresponding database services will be started for each DPU. Otherwise SONiC will only start a global database service. 
Without this new field, we have to add some specific logic for DPU and database service. I don't think it's flexible enough.
https://github.com/sonic-net/sonic-host-services/pull/84